### PR TITLE
Add Dart booster.

### PR DESCRIPTION
**Pull request by marugari**
https://github.com/dmlc/xgboost/pull/1199

### CODE REVIEW

File: src/gbm/gbtree.cc [hunk 1]
```diff
@@ -313,8 +352,9 @@ class GBTree : public GradientBooster {
     }
   }
   // commit new trees all at once
-  inline void CommitModel(std::vector<std::unique_ptr<RegTree> >&& new_trees,
-                          int bst_group) {
+  inline virtual void
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: remove inline, inline and virtual are conflicting with each other

File: doc/parameter.md [hunk 2]
```diff
@@ -74,6 +75,28 @@ Parameters for Tree Booster
...
 
+Additional parameters for Dart
+------------------------------
+* sample_type [default=0]
+  - type of sampling algorithm.
+    - 0: dropped trees are selected uniformly.
+    - 1: dropped trees are selected in proportion to weight.
+* normalize_type [default=0]
+  - type of normalization algorithm.
+    - 0: weight of new trees are 1 / k
```
Comment (tqchen) [inline doc/parameter.md]: update documents, note that user need to pass in 
```
normalize_type="original" from python side, integer will not be accepted
```

File: tests/python/test_basic_models.py [hunk 3]
```diff
@@ -23,6 +23,32 @@ def test_glm(self):
...
+        # specify validations set to watch performance
+        watchlist = [(dtest, 'eval'), (dtrain, 'train')]
+        num_round = 2
+        bst = xgb.train(param, dtrain, num_round, watchlist)
+        # this is prediction
+        preds = bst.predict(dtest, ntree_limit=num_round)
+        labels = dtest.get_label()
+        err = sum(1 for i in range(len(preds)) if int(preds[i] > 0.5) != labels[i]) / float(len(preds))
+        # error must be smaller than 10%
+        assert err < 0.1
```
Comment (tqchen) [inline tests/python/test_basic_models.py]: Let us add a few testcase with the parameters you proposed, e.g. normalize type

File: src/gbm/gbtree.cc [hunk 4]
```diff
@@ -475,14 +515,243 @@ class GBTree : public GradientBooster {
...
+    for (int gid = 0; gid < mparam.num_output_group; ++gid) {
+      this->Pred(inst, -1, gid, root_index, &thread_temp[0],
+                 &(*out_preds)[gid], mparam.num_output_group,
+                 ntree_limit);
+    }
+  }
+
+ protected:
+  // commit new trees all at once
+  inline virtual void
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: remove inline and virtual, add override keyword

File: src/gbm/gbtree.cc [hunk 5]
```diff
@@ -475,14 +515,243 @@ class GBTree : public GradientBooster {
...
+      fo->Write(weight_drop);
+    }
+  }
+
+  // predict the leaf scores with dropout if ntree_limit = 0
+  void Predict(DMatrix* p_fmat,
+               int64_t buffer_offset,
+               std::vector<float>* out_preds,
+               unsigned ntree_limit) override {
+    DropTrees(ntree_limit);
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: is this same as the GBTre except the drop trees? Would be better to leave a comment here

**Comment by tqchen** [comment]
@marugari Thanks for the updates. You donot need to open another PR normally, simply update the contents in your branch will update this PR 

I have made a few more comments. Please also fix the lint error as indicated in https://travis-ci.org/dmlc/xgboost/jobs/132091950

You can reproduce the style check locally by typing make lint

**Comment by tqchen** [comment]
Any updates on this ?

**Comment by marugari** [comment]
Update.
Sorry for my slow-paced work.

File: doc/parameter.md [hunk 6]
```diff
@@ -74,6 +75,28 @@ Parameters for Tree Booster
 * scale_pos_weight, [default=0]
   - Control the balance of positive and negative weights, useful for unbalanced classes. A typical value to consider: sum(negative  cases) / sum(positive cases) See [Parameters Tuning](how_to/param_tuning.md) for more discussion. Also see Higgs Kaggle competition demo for examples: [R](../demo/kaggle-higgs/higgs-train.R ), [py1](../demo/kaggle-higgs/higgs-numpy.py ), [py2](../demo/kaggle-higgs/higgs-cv.py ), [py3](../demo/guide-python/cross_validation.py)
 
+Additional parameters for Dart
+------------------------------
+* sample_type [default=0]
+  - type of sampling algorithm.
+    - "uniform": dropped trees are selected uniformly.
+    - "weighted": dropped trees are selected in proportion to weight.
+* normalize_type [default=0]
```
Comment (tqchen) [inline doc/parameter.md]: default= "original"

**Comment by marugari** [comment]
I fixed it.

**Comment by tqchen** [comment]
One last thing. The name original gbtree and learning rate for normalize
type can be a bit confusing as they are terms used elsewhere. Any idea of
better terms ? What are the terms used by the paper ?

**Comment by marugari** [comment]
In the paper, the authors propose a normalize type. ("original" in my implementation)
It is applicable for learning_rate = 1.
For learning_rate < 1, I add "learning_rate".

I will consider appropriate names.

**Comment by tqchen** [comment]
Does that mean we should automatically choosing the current "learning_rate" by default?

**Comment by marugari** [comment]
If My calculation is correct, we should choose "learning_rate" as normalize type.

**Comment by tqchen** [comment]
Let us directly remove normalize type for now, and go with your implementation by default then :)

**Comment by tqchen** [comment]
please make the update and i will merge the changes in

**Comment by tqchen** [comment]
Thanks this is merged!

**Comment by tqchen** [comment]
@marugari  Thanks for the great job of bringing in the DART trainer.

To make it more well known and widely accessible to users, I would like to invite you to write a markdown guest blogpost introducing what DART is and how you can use it in xgboost with a few code examples. We can post it on the DMLC website as well documents of XGBoost. Please let me know how you think about it and let me know if you need help in reviewing it.

You can first create a PR to https://github.com/dmlc/xgboost/tree/master/doc/tutorials

**Comment by marugari** [comment]
Thank you for your great support.

I'm writing a Japanese blog post. Documents will be submitted afterward.

**Comment by marugari** [comment]
I found my mistakes.
https://github.com/marugari/xgboost/commit/17bafbe47bb5fc8f753e2236714117492bdcb4dc
https://github.com/marugari/Notebooks/blob/master/test_dart.ipynb

I apologize for the trouble.

**Comment by tqchen** [comment]
no problem, please feel free to open a pr to update the code

**Comment by tqchen** [comment]
@marugari Any updates on possible guest blogpost?
